src/domain/core/src/services/compact_service.rs

```diff
@@ -0,0 +1,208 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::path::Path;
+use std::sync::Arc;
+
+use opendatafabric::*;
+use thiserror::Error;
+
+use crate::*;
+
+#[async_trait::async_trait]
+pub trait CompactService: Send + Sync {
+    async fn compact_dataset(
+        &self,
+        dataset_handle: &DatasetHandle,
+        dataset_dir_path: &Path,
+        listener: Option<Arc<dyn CompactionMultiListener>>,
+    ) -> Result<(), CompactError>;
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Errors
+///////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Error)]
+pub enum CompactError {
+    #[error(transparent)]
+    DatasetNotFound(
+        #[from]
+        #[backtrace]
+        DatasetNotFoundError,
+    ),
+    #[error(transparent)]
+    RefNotFound(
+        #[from]
+        #[backtrace]
+        RefNotFoundError,
+    ),
+    #[error(transparent)]
```

Comment:
It feels like this linearized error has no much sense. We should preserve layering here, except for cross-cutting concerns like security.

We asked a high level service to compact the given dataset. What could go wrong:
 - dataset does not exist `DatasetNotFoundError`
 - no access to read or write to this dataset `AccessError`
 - something internal, related to broken lower layers `InternalError`.
 
 The user of `compact_dataset` operates at higher level than the user of individual block operations..

Reply:
Removed